### PR TITLE
Campfire/dashboard required ds

### PR DESCRIPTION
**Pull request by caylahamann**
# Summary

* This adds a new workflow job called `set-dashboard-required-datasources` that flows after `submit-quickstarts`.
* This sets required data sources for each dashboard found in the quickstarts for the PR. 
  * The required data sources are derived from the quickstarts, and the data sources already associated for the published data source

## Pre merge checklist

- [ ] Did you check you NRQL syntax? - Does it work?
- [ ] Did you include an InstallPlan and Documentation reference?
- [ ] Are all documentation links publically accessible?
- [ ] Did you check your descriptive content for [voice and tone](https://docs.newrelic.com/docs/style-guide/writing-strategies/voice-strategies-docs-sound-new-relic/)? 
- [ ] Did you check your descriptive content for spelling and grammar errors?

### Dashboards 

- [ ] Does the PR contain a screenshot for each of your dashboards?
- [ ] Do your screenshots show data?
- [ ] Has the [sanitization script](https://github.com/newrelic/newrelic-quickstarts/blob/main/CONTRIBUTING.md#dashboards) been run on each dashboard?

### Alerts

- [ ] Did you check that your alerts actually work?

### CODE REVIEW

File: utils/lib/Dashboard.ts [hunk 1]
```diff
@@ -106,6 +145,65 @@ class Dashboard extends Component<DashboardConfig, QuickstartDashboardInput> {
...
+    const { data, errors } = await fetchNRGraphqlResults<
+      DashboardRequiredDataSourcesQueryVariables,
+      DashboardRequiredDataSourcesQueryResults
+    >({
+      queryString: DASHBOARD_REQUIRED_DATA_SOURCES_QUERY,
+      variables: { id: templateId },
+    });
+
+    if (errors) {
+      Promise.reject(`Query for dashboard template id failed`);
```
Comment (caylahamann) [inline utils/lib/Dashboard.ts:165]: If we reject the promise here, don't we never return the errors from query for the data sources? @MichelLosier 

So either we need to get rid of line 188-190 or if we only want to reject on the outer layer then we can get rid of this and instead just rely on the return of line 188-190

Comment (caylahamann) [reply utils/lib/Dashboard.ts:165]: I'm go ahead of getting rid of those lines and just rely on rejecting the promise here

Comment (MichelLosier) [reply utils/lib/Dashboard.ts]: ya good call, went ahead and removed that promise rejection

File: utils/lib/Dashboard.ts [hunk 2]
```diff
@@ -106,6 +152,61 @@ class Dashboard extends Component<DashboardConfig, QuickstartDashboardInput> {
...
+      variables: { id: templateId },
+    });
+
+    const ids =
+      data?.actor?.nr1Catalog?.dashboardTemplate?.metadata?.requiredDataSources?.map(
+        ({ id }) => id
+      );
+
+    return { ids: ids ?? [], errors };
+  }
```
Comment (josephgregoryii) [inline utils/lib/Dashboard.ts:177]:
```suggestion
    const ids =
      data?.actor?.nr1Catalog?.dashboardTemplate?.metadata?.requiredDataSources?.map(
        ({ id }) => id
      ) ?? [];

    return { ids, errors };
  }
```

File: utils/lib/__tests__/Dashboard.test.js [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
 import * as fs from 'fs';
+import * as nrGraphqlHelpers from '../nr-graphql-helpers';
 
-import { GITHUB_REPO_BASE_URL } from '../../constants';
+import {
+  GITHUB_REPO_BASE_URL,
+  DASHBOARD_REQUIRED_DATA_SOURCES_QUERY,
+  DASHBOARD_SET_REQUIRED_DATA_SOURCES_MUTATION,
+} from '../../constants';
+import { nil } from 'ajv';
```
Comment (josephgregoryii) [inline utils/lib/__tests__/Dashboard.test.js:11]: I'm not sure if we need this import - i'm not seeing it used?

File: utils/lib/nr-graphql-helpers.ts [hunk 4]
```diff
@@ -208,6 +225,83 @@ export const getPublishedDataSourceIds =
     return { coreDataSourceIds, errors };
   };
 
+// Lets put these types in own file in util/types
```
Comment (josephgregoryii) [inline utils/lib/nr-graphql-helpers.ts:228]: I think we were meaning to move these into their own types file.

**Review by josephgregoryii — COMMENTED** [review]
Looks great! just a few comments below. Some can be ignored

**Review by josephgregoryii — COMMENTED** [review]
Looks good! Approved, waiting to give git approval for after the code freeze

**Review by aswanson-nr — APPROVED** [review]

**Pull request merged**